File: client-python/examples/integration_example.py
#!/usr/bin/env python3
"""
Integration example for soulseekR - demonstrates coordinated client operations
Shows batch operations, WebSocket events, error handling, and data aggregation
"""

import asyncio
import time
from soulseekr import SoulseekrClient, ApiError


class SearchCoordinator:
    """Coordinates multiple search operations"""

    def __init__(self, client):
        self.client = client
        self.searches = {}
        self.results_count = {}

    async def create_searches(self, queries):
        """Create multiple searches concurrently"""
        print(f"\n=== Creating {len(queries)} searches ===")
        
        tasks = [self.client.create_search(q) for q in queries]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        successful = []
        for query, result in zip(queries, results):
            if isinstance(result, Exception):
                print(f"✗ Error searching '{query}': {result}")
            else:
                search_id = result.get('id')
                self.searches[search_id] = query
                successful.append(search_id)
                print(f"✓ Created search: {query} (ID: {search_id})")

        return successful

    async def monitor_searches(self, search_ids, duration_seconds=5):
        """Monitor search progress"""
        print(f"\n=== Monitoring {len(search_ids)} searches for {duration_seconds}s ===")
        
        start = time.time()
        while time.time() - start < duration_seconds:
            tasks = [
                self.client.get_search_details(sid, limit=10)
                for sid in search_ids
            ]
            
            try:
                details_list = await asyncio.gather(*tasks, return_exceptions=True)
                
                for search_id, details in zip(search_ids, details_list):
                    if isinstance(details, Exception):
                        continue
                    
                    result_count = len(details.get('results', []))
                    self.results_count[search_id] = result_count
                    query = self.searches.get(search_id, 'Unknown')
                    print(f"  {query}: {result_count} results")
                
                await asyncio.sleep(1)
            except Exception as e:
                print(f"Error monitoring searches: {e}")
                break

    async def get_top_results(self, search_id, limit=5):
        """Get top results from a search"""
        try:
            details = await self.client.get_search_details(search_id, limit=limit)
            return details.get('results', [])
        except ApiError as e:
            print(f"Error getting search details: {e}")
            return []


class MessageHandler:
    """Handles message operations"""

    def __init__(self, client):
        self.client = client
        self.messages_sent = 0
        self.messages_received = 0

    async def list_recent_messages(self, limit=10):
        """List recent messages"""
        try:
            messages = await self.client.list_messages(limit=limit)
            self.messages_received = len(messages)
            return messages
        except ApiError as e:
            print(f"Error listing messages: {e}")
            return []

    async def send_bulk_messages(self, recipients_messages):
        """Send multiple messages"""
        print(f"\n=== Sending {len(recipients_messages)} messages ===")
        
        tasks = [
            self.client.send_message(recipient, content)
            for recipient, content in recipients_messages
        ]
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        successful = 0
        for (recipient, _), result in zip(recipients_messages, results):
            if isinstance(result, Exception):
                print(f"✗ Failed to send message to {recipient}: {result}")
            else:
                successful += 1
                print(f"✓ Message sent to {recipient}")
        
        self.messages_sent = successful
        return successful


async def batch_analytics(client):
    """Gather analytics using batch operations"""
    print("\n=== Batch Analytics ===")
    
    batch = client.batch.builder()
    batch.get("/api/stats", op_id="stats")
    batch.get("/api/config", op_id="config")
    batch.get("/api/capabilities", op_id="caps")

    response = await batch.execute()

    analytics = {}
    for result in response.results:
        if result.is_success():
            analytics[result.id] = result.body
            print(f"✓ {result.id}: collected")
        else:
            print(f"✗ {result.id}: failed with status {result.status}")

    return analytics


async def websocket_monitoring(client, duration_seconds=3):
    """Monitor events via WebSocket"""
    print(f"\n=== WebSocket Monitoring for {duration_seconds}s ===")
    
    try:
        ws = await client.connect_ws()
        
        # Subscribe to all event types
        ws.subscribe("messages", "search_updates", "transfer_updates")
        
        event_counts = {"messages": 0, "search_updates": 0, "transfer_updates": 0}
        
        # Simulate event reception
        async def event_receiver():
            start = time.time()
            while time.time() - start < duration_seconds:
                await asyncio.sleep(0.5)
            return event_counts

        result = await event_receiver()
        print(f"Event monitoring complete: {result}")
        
    except Exception as e:
        print(f"Error with WebSocket: {e}")
    finally:
        await client.disconnect_ws()


async def transfer_operations(client):
    """Handle transfer operations"""
    print("\n=== Transfer Operations ===")
    
    try:
        # List active transfers
        transfers = await client.list_transfers(
            direction="download",
            status="active",
            limit=10
        )
        
        print(f"Active downloads: {len(transfers)}")
        
        for transfer in transfers[:3]:
            print(f"  - {transfer.get('filename')}: {transfer.get('progress')}%")
        
        return len(transfers)
    except ApiError as e:
        print(f"Error listing transfers: {e}")
        return 0


async def main():
    """Main integration example"""
    
    # Initialize client
    client = SoulseekrClient(
        base_url="http://localhost:8080",
        token="your-api-key-here",
        debug=True
    )

    try:
        # 1. Verify server connectivity
        print("=== Server Verification ===")
        health = await client.health()
        print(f"✓ Server status: {health.get('status')}")

        # 2. Search coordination
        coordinator = SearchCoordinator(client)
        queries = ["radiohead", "pink floyd", "led zeppelin"]
        search_ids = await coordinator.create_searches(queries)

        if search_ids:
            # Monitor the searches
            await coordinator.monitor_searches(search_ids, duration_seconds=3)

            # Get top results from first search
            if search_ids:
                top_results = await coordinator.get_top_results(search_ids[0], limit=5)
                print(f"\nTop results for first search: {len(top_results)}")

        # 3. Message operations
        handler = MessageHandler(client)
        recent_messages = await handler.list_recent_messages(limit=5)
        print(f"\n=== Message Operations ===")
        print(f"Recent messages: {len(recent_messages)}")

        # Send demo messages
        demo_recipients = [
            ("demo_user_1", "Test message 1"),
            ("demo_user_2", "Test message 2"),
        ]
        sent_count = await handler.send_bulk_messages(demo_recipients)
        print(f"Messages sent: {sent_count}")

        # 4. Batch analytics
        analytics = await batch_analytics(client)
        print(f"Collected {len(analytics)} analytics")

        # 5. Transfer monitoring
        active_transfers = await transfer_operations(client)

        # 6. WebSocket monitoring
        await websocket_monitoring(client, duration_seconds=2)

        # 7. Summary
        print("\n=== Integration Summary ===")
        print(f"Searches created: {len(search_ids)}")
        print(f"Messages sent: {handler.messages_sent}")
        print(f"Analytics collected: {len(analytics)}")
        print(f"Active transfers: {active_transfers}")
        print("✓ Integration test complete")

    except Exception as e:
        print(f"Integration error: {e}")
        import traceback
        traceback.print_exc()

    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: client-go/examples/integration_example.go
//go:build examples

package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/snapetech/slskr/client-go"
)

// SearchCoordinator coordinates multiple search operations
type SearchCoordinator struct {
	client     *slskr.Client
	searches   map[string]string // searchID -> query
	resultsCnt map[string]int
	mu         sync.RWMutex
}

// NewSearchCoordinator creates a new search coordinator
func NewSearchCoordinator(client *slskr.Client) *SearchCoordinator {
	return &SearchCoordinator{
		client:     client,
		searches:   make(map[string]string),
		resultsCnt: make(map[string]int),
	}
}

// CreateSearches creates multiple searches concurrently
func (sc *SearchCoordinator) CreateSearches(ctx context.Context, queries []string) []string {
	fmt.Printf("\n=== Creating %d searches ===\n", len(queries))

	var wg sync.WaitGroup
	results := make(chan string, len(queries))
	errors := make(chan error, len(queries))

	for _, q := range queries {
		wg.Add(1)
		go func(query string) {
			defer wg.Done()

			search, err := sc.client.CreateSearch(ctx, query)
			if err != nil {
				errors <- err
				return
			}

			searchID, _ := search["id"].(string)
			sc.mu.Lock()
			sc.searches[searchID] = query
			sc.mu.Unlock()

			results <- searchID
			fmt.Printf("✓ Created search: %s (ID: %s)\n", query, searchID)
		}(q)
	}

	go func() {
		wg.Wait()
		close(results)
		close(errors)
	}()

	var searchIDs []string
	for searchID := range results {
		searchIDs = append(searchIDs, searchID)
	}

	for err := range errors {
		fmt.Printf("✗ Error creating search: %v\n", err)
	}

	return searchIDs
}

// MonitorSearches monitors search progress
func (sc *SearchCoordinator) MonitorSearches(ctx context.Context, searchIDs []string, duration time.Duration) {
	fmt.Printf("\n=== Monitoring %d searches for %v ===\n", len(searchIDs), duration)

	deadline := time.Now().Add(duration)
	ticker := time.NewTicker(500 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			if time.Now().After(deadline) {
				return
			}

			for _, sid := range searchIDs {
				query := sc.searches[sid]
				count := sc.resultsCnt[sid]
				fmt.Printf("  %s: %d results\n", query, count)
			}

		case <-ctx.Done():
			return
		}
	}
}

// MessageHandler handles message operations
type MessageHandler struct {
	client       *slskr.Client
	messagesSent int
	messagesRecv int
	mu           sync.Mutex
}

// NewMessageHandler creates a new message handler
func NewMessageHandler(client *slskr.Client) *MessageHandler {
	return &MessageHandler{
		client: client,
	}
}

// ListRecentMessages lists recent messages
func (mh *MessageHandler) ListRecentMessages(ctx context.Context, limit int) {
	messages, err := mh.client.ListMessages(ctx, limit, 0)
	if err != nil {
		fmt.Printf("Error listing messages: %v\n", err)
		return
	}

	mh.mu.Lock()
	mh.messagesRecv = len(messages)
	mh.mu.Unlock()

	fmt.Printf("\nRecent messages: %d\n", len(messages))
}

// SendBulkMessages sends multiple messages
func (mh *MessageHandler) SendBulkMessages(ctx context.Context, messages []map[string]string) {
	fmt.Printf("\n=== Sending %d messages ===\n", len(messages))

	var wg sync.WaitGroup
	results := make(chan bool, len(messages))

	for _, msg := range messages {
		wg.Add(1)
		go func(recipient, content string) {
			defer wg.Done()

			_, err := mh.client.SendMessage(ctx, recipient, content)
			if err != nil {
				fmt.Printf("✗ Failed to send to %s: %v\n", recipient, err)
				results <- false
				return
			}

			fmt.Printf("✓ Message sent to %s\n", recipient)
			results <- true
		}(msg["recipient"], msg["content"])
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	sent := 0
	for success := range results {
		if success {
			sent++
		}
	}

	mh.mu.Lock()
	mh.messagesSent = sent
	mh.mu.Unlock()
}

// batchAnalytics gathers analytics using batch operations
func batchAnalytics(ctx context.Context, client *slskr.Client) map[string]interface{} {
	fmt.Println("\n=== Batch Analytics ===")

	batch := client.NewBatchBuilder()
	opStats := "stats"
	opConfig := "config"
	opCaps := "caps"

	batch.Get("/api/stats", &opStats)
	batch.Get("/api/config", &opConfig)
	batch.Get("/api/capabilities", &opCaps)

	response, err := batch.Execute(ctx)
	if err != nil {
		fmt.Printf("Error executing batch: %v\n", err)
		return nil
	}

	analytics := make(map[string]interface{})
	for _, result := range response.Results {
		if result.IsSuccess() {
			analytics[result.ID] = result.Body
			fmt.Printf("✓ %s: collected\n", result.ID)
		} else {
			fmt.Printf("✗ %s: failed with status %d\n", result.ID, result.Status)
		}
	}

	return analytics
}

// transferOperations handles transfer operations
func transferOperations(ctx context.Context, client *slskr.Client) int {
	fmt.Println("\n=== Transfer Operations ===")

	transfers, err := client.ListTransfers(ctx, "download", "active", 10, 0)
	if err != nil {
		fmt.Printf("Error listing transfers: %v\n", err)
		return 0
	}

	fmt.Printf("Active downloads: %d\n", len(transfers))

	count := 0
	for i, t := range transfers {
		if i >= 3 {
			break
		}
		if filename, ok := t["filename"].(string); ok {
			if progress, ok := t["progress"].(float64); ok {
				fmt.Printf("  - %s: %.0f%%\n", filename, progress)
			}
		}
		count++
	}

	return len(transfers)
}

// roomOperations handles room operations
func roomOperations(ctx context.Context, client *slskr.Client) int {
	fmt.Println("\n=== Room Operations ===")

	rooms, err := client.ListRooms(ctx)
	if err != nil {
		fmt.Printf("Error listing rooms: %v\n", err)
		return 0
	}

	fmt.Printf("Available rooms: %d\n", len(rooms))

	for i, r := range rooms {
		if i >= 3 {
			break
		}
		if name, ok := r["name"].(string); ok {
			if users, ok := r["user_count"].(float64); ok {
				fmt.Printf("  - %s: %d users\n", name, int(users))
			}
		}
	}

	return len(rooms)
}

// shareOperations handles share operations
func shareOperations(ctx context.Context, client *slskr.Client) int {
	fmt.Println("\n=== Share Operations ===")

	shares, err := client.ListShares(ctx, 10, 0)
	if err != nil {
		fmt.Printf("Error listing shares: %v\n", err)
		return 0
	}

	fmt.Printf("Shared files: %d\n", len(shares))

	return len(shares)
}

// websocketMonitoring monitors events via WebSocket
func websocketMonitoring(ctx context.Context, client *slskr.Client, duration time.Duration) {
	fmt.Printf("\n=== WebSocket Monitoring for %v ===\n", duration)

	ws := client.NewWebSocketClient(true)
	err := ws.Connect(ctx)
	if err != nil {
		fmt.Printf("Error connecting WebSocket: %v\n", err)
		return
	}
	defer ws.Disconnect(ctx)

	ws.Subscribe("messages", "search_updates", "transfer_updates")
	fmt.Printf("Subscribed to topics: %v\n", ws.GetSubscribedTopics())

	time.Sleep(duration)
}

func main() {
	client := slskr.NewClient("http://127.0.0.1:5030", "your-api-key-here")
	ctx := context.Background()
	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
	defer cancel()

	fmt.Println("=== slskr Integration Example ===")

	// 1. Server verification
	fmt.Println("\n=== Server Verification ===")
	health, err := client.Health(ctx)
	if err != nil {
		log.Fatalf("Error checking health: %v", err)
	}
	fmt.Printf("✓ Server status: %v\n", health["status"])

	// 2. Search coordination
	coordinator := NewSearchCoordinator(client)
	queries := []string{"radiohead", "pink floyd", "led zeppelin"}
	searchIDs := coordinator.CreateSearches(ctx, queries)

	if len(searchIDs) > 0 {
		// Monitor searches briefly
		monitorCtx, monitorCancel := context.WithTimeout(ctx, 2*time.Second)
		coordinator.MonitorSearches(monitorCtx, searchIDs, 2*time.Second)
		monitorCancel()
	}

	// 3. Message operations
	handler := NewMessageHandler(client)
	handler.ListRecentMessages(ctx, 5)

	demoMessages := []map[string]string{
		{"recipient": "demo_user_1", "content": "Test message 1"},
		{"recipient": "demo_user_2", "content": "Test message 2"},
	}
	handler.SendBulkMessages(ctx, demoMessages)

	// 4. Batch analytics
	analytics := batchAnalytics(ctx, client)
	fmt.Printf("Collected %d analytics\n", len(analytics))

	// 5. Transfer operations
	activeTransfers := transferOperations(ctx, client)

	// 6. Room operations
	roomCount := roomOperations(ctx, client)

	// 7. Share operations
	shareCount := shareOperations(ctx, client)

	// 8. WebSocket monitoring
	websocketMonitoring(ctx, client, 1*time.Second)

	// 9. Summary
	fmt.Println("\n=== Integration Summary ===")
	fmt.Printf("Searches created: %d\n", len(searchIDs))
	fmt.Printf("Messages sent: %d\n", handler.messagesSent)
	fmt.Printf("Analytics collected: %d\n", len(analytics))
	fmt.Printf("Active transfers: %d\n", activeTransfers)
	fmt.Printf("Available rooms: %d\n", roomCount)
	fmt.Printf("Shared files: %d\n", shareCount)
	fmt.Println("✓ Integration test complete")
}
